doc/user/content/sql/subscribe.md

```diff
@@ -517,6 +522,62 @@ When you're done, you can drop the `counter` load generator source:
 DROP SOURCE counter;
 ```
 
+### Durable subscriptions
+
+Because `SUBSCRIBE` requests happen over the network, these connections might
+get disrupted for both expected and unexpected reasons. You can adjust the
+[history retention period](/transform-data/patterns/time-travel-queries/#history-retention-period) for
+the objects a subscription depends on, and then use [`AS OF`](#as-of) to pick
+up where you left off on connection drops — this ensures that no data is lost
+in the subscription process, and avoids the need for re-snapshotting the data.
+
+To set up a durable subscription in your application:
+
+1. In Materialize, configure the history retention period for the object
+(s) queried in the `SUBSCRIBE`. Choose a duration you expect will allow you to
+recover in case of connection drops. One hour (`1h`) is a good place to start,
+though you should be mindful of the [resource utilization impact](/transform-data/patterns/time-travel-queries/#resource-utilization)
+of increasing an object's history retention period.
+
+1. In order to restart your application without losing or re-snapshotting data
+after a connection drop, you need to store the latest timestamp processed for
+the subscription (either in Materialize, or elsewhere in your application
+state). This will allow you to resume using the retained history upstream.
+
+1. The first time you start the subscription, run the following
+continuous query against Materialize in your application code:
+
+   ```sql
+   SUBSCRIBE (<your query>) WITH (PROGRESS, SNAPSHOT true);
+   ```
+
+   If you do not need a full snapshot to bootstrap your application,  change
+   this to `SNAPSHOT false`.
+
+1. As results come in continuously, buffer the latest results in memory until you receive
+a [progress](#progress) message. At that point, the data up until the progress message
+is complete, so you can:
+
+   1. Process all the buffered data in your application.
+   1. Persist the `mz_timestamp` of the progress message.
```

Comment:
This description is good, but leaves out some possibly useful details for users to get all the edge cases correct. What if their application gets a progress message and has processed some of the buffered data, and then it crashes? These data will then be present in their next batch because they will use an old AS OF. Do we want to mention idempotency here? Or maybe we should recommend that when the progress message comes, the buffered data AND the timestamp are both committed to some database (mz or theirs) in a single transaction. Then they can process data out of that storage and even if their app goes down, they know (nearly) exactly where they left off and can (almost) avoid duplicating any work. The nearly and almost there don't apply to the one exact message that was being processed at the time of crash (they would need some other transaction or idempotency guarantee to fix that).

I think the take away is: the timestamp and buffered data should get written down transactionally together.

Reply:
How can they get written down transactionally together if the place where the buffered data is written out is different from where they store the timestamp though? E.g. if they're pushing user notifications based on the buffered data, they'd have to use some other datastore for the timestamp.

Reply:
they cannot! but that's the only way to do it correctly!

Reply:
see for example google app engine tasks queues that "enqueue a task as part of a Datastore transaction" enabling exactly this. i'm aware this api isn't available to, well, pretty much everyone, but it is definitely a thing.

i have other ideas but am out of time atm. we can also brainstorm stuff together.

https://cloud.google.com/appengine/docs/legacy/standard/python/taskqueue/push/creating-tasks#enqueuing_tasks_in_cloud_datastore_transactions

Reply:
I just pushed up a change where I added a section on idempotency that covers what happens in crash. I might iterate on it a bit more, but it's at least worth calling out. Thanks for sharing!